### PR TITLE
SHA-256 Stream Clear API to call lower level mbedtls free

### LINKED ISSUE #31569: CHIP Crypto PAL MBEDTLS implementation inconsistent in how it cleans SHA-256 resources.
When using HW supported Crypto accelerators for SHA-256 operations the CHIPCryptoPALmbedTLS implementation does not call the low level mbedtls_sha256_free() API when cleaning up its memory during new CASE sessions. 
This can be seen in the CASE session initialization here:
https://github.com/project-chip/connectedhomeip/blob/master/src/protocols/secure_channel/CASESession.cpp#L445

As only a memory clear is done, there is no way for lower level HW implementations to know that this resource is no longer needed, leaving HW implementations in an unknown state. 
 
The SHA-256 Clear() API should align to the others in this file, to first call the low level mbedtls_sha256_free() *alt API prior to wiping out the class buffer.
  
For example:
Within the Clear() implementation for SHA-256, we only memset the buffer so 0.
https://github.com/project-chip/connectedhomeip/blob/master/src/crypto/CHIPCryptoPALmbedTLS.cpp#L273
 
For the remaining Crypto calls within this file, however, the Clear() APIs correctly call the lower level *free() implementations:
https://github.com/project-chip/connectedhomeip/blob/master/src/crypto/CHIPCryptoPALmbedTLS.cpp#L774
 
https://github.com/project-chip/connectedhomeip/blob/master/src/crypto/CHIPCryptoPALmbedTLS.cpp#L917

### PR DESCRIPTION
**Pull request by adabreuti**
Fixes #31569 by updating the Clear() implementation for Hash_SHA256_stream to call the lower mbedtls free API prior to clearing the buffer.

Validated on the TI platform with back-to-back CASE session establishments with HW accelerated SHA-256 enabled & disabled.

### CODE REVIEW

**Comment by adabreuti** [comment]
@andreilitvin @bzbarsky-apple Is third party validation necessary prior to approval for a PR which updates a generic Crypto API? If so, what are the routes for requesting such validation?

File: third_party/ti_simplelink_sdk/repo_cc13xx_cc26xx
Comment (andy31415) [inline third_party/ti_simplelink_sdk/repo_cc13xx_cc26xx:1]: This change may not be intentional ...

Comment (andy31415) [reply third_party/ti_simplelink_sdk/repo_cc13xx_cc26xx:1]: or at least does not seem directly related with mbedTLS changes

Comment (adabreuti) [reply third_party/ti_simplelink_sdk/repo_cc13xx_cc26xx:1]: With this mbedTLS fix, we are able to properly use our Crypto API's. Should this be split into a separate PR instead?

Comment (bzbarsky-apple) [reply third_party/ti_simplelink_sdk/repo_cc13xx_cc26xx:1]: Yes, it should.

**Comment by bzbarsky-apple** [comment]
@adabreuti Mostly this needs review from someone who knows something about the mbedTLS bits....

File: third_party/ti_simplelink_sdk/repo_cc13xx_cc26xx
Comment (adabreuti) [reply third_party/ti_simplelink_sdk/repo_cc13xx_cc26xx:1]: Sounds good, the commit has been reverted.

**Review by hawk248 — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Pull request merged**